apps/wing-console/console/server/src/utils/test-runner/test-runner.ts

```diff
@@ -0,0 +1,259 @@
+  try {
+    const t = await client.runTest(resourcePath);
+    for (const log of t.traces) {
+      if (log.type === TraceType.LOG) {
+        logger.log(log.data.message, "simulator");
+      } else {
+        logger.verbose(log.data.message, "simulator");
+      }
+    }
+    result = {
+      ...result,
+      ...t,
+    };
+    if (result.error) {
+      throw new Error(result.error);
+    }
+    logger.log(
+      `Test "${getTestName(resourcePath)}" succeeded (${
+        Date.now() - startTime
+      }ms)`,
+      "console",
+      {
+        messageType: "success",
+      },
+    );
+  } catch (error: any) {
+    let output = await formatTraceError(error?.message);
+    logger.log(output, "console", {
+      messageType: "fail",
+    });
+    logger.log(
+      `Test "${getTestName(resourcePath)}" failed (${
+        Date.now() - startTime
+      }ms)`,
+      "console",
+      {
+        messageType: "fail",
+      },
+    );
+  }
+  result.time = Date.now() - startTime;
+  return result;
+};
+
+export const createTestRunner = ({
+  wingfile,
+  platform,
+  watchGlobs,
+  logger,
+}: CreateTestRunnerProps): TestRunner => {
+  const testsState = createTestStateManager();
+
+  const simulatorManager = createSimulatorManager({
+    wingfile,
+    platform,
+    watchGlobs,
+  });
+
+  const onTestChangeCallbacks: Array<() => void> = [];
+
+  const runOnTestChangeCallbacks = () => {
+    for (const callback of onTestChangeCallbacks) {
+      callback();
+    }
+  };
+
+  const initialize = async () => {
+    testsState.restart();
+
+    const simulator = await simulatorManager.getSimulator();
+
+    const simTestRunner = simulator.getResource(
+      "root/cloud.TestRunner",
+    ) as ITestRunnerClient;
+
+    const tests = await simTestRunner.listTests();
+
+    simulator.stop();
+
+    testsState.setTests(
+      tests.map((test) => ({
+        id: test,
+        label: getTestName(test),
+        status: "idle",
+        datetime: Date.now(),
+      })),
+    );
+
+    runOnTestChangeCallbacks();
+  };
+
+  const status = (): TestStatus => {
+    if (!testsState.initialized()) {
+      return "uninitialized";
+    }
+    if (testsState.getTests().some((t) => t.status === "running")) {
+      return "running";
+    }
+    if (testsState.getTests().some((t) => t.status === "error")) {
+      return "error";
+    }
+    if (testsState.getTests().some((t) => t.status === "idle")) {
+      return "idle";
+    }
+    return "success";
+  };
+
+  const runTest = async (testId: string) => {
+    const simulator = await simulatorManager.getSimulator();
+
+    const response = await executeTest(simulator, testId, logger);
+    testsState.setTest({
+      id: testId,
+      label: getTestName(testId),
+      status: response.error ? "error" : "success",
+      time: response.time,
+      datetime: Date.now(),
+    });
+
+    runOnTestChangeCallbacks();
+    simulator.stop();
+  };
+
+  const runAllTests = async () => {
```

Comment:
Calling `runTest` and `runAllTests` in sequence should work without crashing. I suggest that both of these methods use some kind of shared test queue.